# Projet DATA Partie 1&2

Voici mon projet data sur le dataset "heart_disease_data.csv" avant tout chose je vais analyser et explorer mon dataset afin de de ressortir les valeurs importantes et de nettoyer les données. Suite à cela on va utiliser des methodes vu en cours pour faire une visualisation des données (Voir la Partie 3.0)

## Afficher le base de données

Je vais d'abord "import" la bibliotheque panda.

In [1]:
import pandas as pd
data = pd.read_csv('heart_disease_data.csv')

In [2]:
data.info()
# J'affiche les informations sur la base de données pour voir les types de données et les valeurs manquantes

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 918 entries, 0 to 917
Data columns (total 12 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Age             918 non-null    int64  
 1   Sex             918 non-null    object 
 2   ChestPainType   918 non-null    object 
 3   RestingBP       918 non-null    int64  
 4   Cholesterol     918 non-null    int64  
 5   FastingBS       918 non-null    int64  
 6   RestingECG      918 non-null    object 
 7   MaxHR           918 non-null    int64  
 8   ExerciseAngina  918 non-null    object 
 9   Oldpeak         918 non-null    float64
 10  ST_Slope        918 non-null    object 
 11  HeartDisease    918 non-null    int64  
dtypes: float64(1), int64(6), object(5)
memory usage: 86.2+ KB


Il y'a 918 lignes et 12 columnes.

On remarque qu'il n'y a aucune données nulles dans les lignes de la bases de données.
Je vais afficher les premieres lignes et les dernières lignes pour voir si il y'a des doublons visibles.

In [3]:
data.head()

Age Sex ChestPainType  RestingBP  Cholesterol  FastingBS RestingECG  MaxHR  \
0   40   M           ATA        140          289          0     Normal    172   
1   49   F           NAP        160          180          0     Normal    156   
2   37   M           ATA        130          283          0         ST     98   
3   48   F           ASY        138          214          0     Normal    108   
4   54   M           NAP        150          195          0     Normal    122   

  ExerciseAngina  Oldpeak ST_Slope  HeartDisease  
0              N      0.0       Up             0  
1              N      1.0     Flat             1  
2              N      0.0       Up             0  
3              Y      1.5     Flat             1  
4              N      0.0       Up             0

In [4]:
data.tail()

Age Sex ChestPainType  RestingBP  Cholesterol  FastingBS RestingECG  \
913   45   M            TA        110          264          0     Normal   
914   68   M           ASY        144          193          1     Normal   
915   57   M           ASY        130          131          0     Normal   
916   57   F           ATA        130          236          0        LVH   
917   38   M           NAP        138          175          0     Normal   

     MaxHR ExerciseAngina  Oldpeak ST_Slope  HeartDisease  
913    132              N      1.2     Flat             1  
914    141              N      3.4     Flat             1  
915    115              Y      1.2     Flat             1  
916    174              N      0.0     Flat             1  
917    173              N      0.0       Up             0

Par mesure de precaution je vais retirer les doublons dans la base de données.

In [5]:
data.drop_duplicates(inplace=True)
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 918 entries, 0 to 917
Data columns (total 12 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Age             918 non-null    int64  
 1   Sex             918 non-null    object 
 2   ChestPainType   918 non-null    object 
 3   RestingBP       918 non-null    int64  
 4   Cholesterol     918 non-null    int64  
 5   FastingBS       918 non-null    int64  
 6   RestingECG      918 non-null    object 
 7   MaxHR           918 non-null    int64  
 8   ExerciseAngina  918 non-null    object 
 9   Oldpeak         918 non-null    float64
 10  ST_Slope        918 non-null    object 
 11  HeartDisease    918 non-null    int64  
dtypes: float64(1), int64(6), object(5)
memory usage: 86.2+ KB


Remarque : Les doublons on ete supprimés sauf qu'il n'y en avait aucun. On peut donc dire que la base est completement nettoyer.

## Preparation des données

Je vais maintenant transformer les données non numériques en données numériques pour faciliter l'utilisation des algorithmes dans les futures parties.

### Sex

In [6]:
data['Sex'] = data['Sex'].map({'M' : 0, 'F' : 1})
data['Sex']

0      0
1      1
2      0
3      1
4      0
      ..
913    0
914    0
915    0
916    1
917    0
Name: Sex, Length: 918, dtype: int64

### ChestPainType

Dans ce cas, vu qu'il y'a plusieurs valeurs on va cree des colonnes pour chaques valeurs puis supprimer la colonne 'ChestPainType'

In [7]:
CPL = ['TA', 'ATA', 'NAP', 'ASY']

for name in CPL:
    data[name + "_ChestPain"] = (data['ChestPainType'] == name).astype(int)

data = data.drop('ChestPainType', axis=1)

data.info()



<class 'pandas.core.frame.DataFrame'>
RangeIndex: 918 entries, 0 to 917
Data columns (total 15 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Age             918 non-null    int64  
 1   Sex             918 non-null    int64  
 2   RestingBP       918 non-null    int64  
 3   Cholesterol     918 non-null    int64  
 4   FastingBS       918 non-null    int64  
 5   RestingECG      918 non-null    object 
 6   MaxHR           918 non-null    int64  
 7   ExerciseAngina  918 non-null    object 
 8   Oldpeak         918 non-null    float64
 9   ST_Slope        918 non-null    object 
 10  HeartDisease    918 non-null    int64  
 11  TA_ChestPain    918 non-null    int32  
 12  ATA_ChestPain   918 non-null    int32  
 13  NAP_ChestPain   918 non-null    int32  
 14  ASY_ChestPain   918 non-null    int32  
dtypes: float64(1), int32(4), int64(7), object(3)
memory usage: 93.4+ KB


### RestingECG

In [8]:

RECG = ['Normal', 'ST', 'LVH']
for name in RECG:
    data[name + "_RestingECG"] = (data['RestingECG'] == name).astype(int)

data = data.drop('RestingECG', axis=1)


data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 918 entries, 0 to 917
Data columns (total 17 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Age                918 non-null    int64  
 1   Sex                918 non-null    int64  
 2   RestingBP          918 non-null    int64  
 3   Cholesterol        918 non-null    int64  
 4   FastingBS          918 non-null    int64  
 5   MaxHR              918 non-null    int64  
 6   ExerciseAngina     918 non-null    object 
 7   Oldpeak            918 non-null    float64
 8   ST_Slope           918 non-null    object 
 9   HeartDisease       918 non-null    int64  
 10  TA_ChestPain       918 non-null    int32  
 11  ATA_ChestPain      918 non-null    int32  
 12  NAP_ChestPain      918 non-null    int32  
 13  ASY_ChestPain      918 non-null    int32  
 14  Normal_RestingECG  918 non-null    int32  
 15  ST_RestingECG      918 non-null    int32  
 16  LVH_RestingECG     918 non

### ExerciseAngina

In [9]:
data['ExerciseAngina'] = data['ExerciseAngina'].map({'N' : 0, 'Y' : 1})
data['ExerciseAngina']

0      0
1      0
2      0
3      1
4      0
      ..
913    0
914    0
915    1
916    0
917    0
Name: ExerciseAngina, Length: 918, dtype: int64

### ST_Slope

In [10]:
STS = ['Up', 'Flat', 'Down']
for name in STS:
    data[name + "_ST_Slope"] = (data['ST_Slope'] == name).astype(int)
data = data.drop('ST_Slope', axis=1)
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 918 entries, 0 to 917
Data columns (total 19 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Age                918 non-null    int64  
 1   Sex                918 non-null    int64  
 2   RestingBP          918 non-null    int64  
 3   Cholesterol        918 non-null    int64  
 4   FastingBS          918 non-null    int64  
 5   MaxHR              918 non-null    int64  
 6   ExerciseAngina     918 non-null    int64  
 7   Oldpeak            918 non-null    float64
 8   HeartDisease       918 non-null    int64  
 9   TA_ChestPain       918 non-null    int32  
 10  ATA_ChestPain      918 non-null    int32  
 11  NAP_ChestPain      918 non-null    int32  
 12  ASY_ChestPain      918 non-null    int32  
 13  Normal_RestingECG  918 non-null    int32  
 14  ST_RestingECG      918 non-null    int32  
 15  LVH_RestingECG     918 non-null    int32  
 16  Up_ST_Slope        918 non

On affiche la nouvelle base de données pour bien verifier tous les changements.

In [11]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 918 entries, 0 to 917
Data columns (total 19 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Age                918 non-null    int64  
 1   Sex                918 non-null    int64  
 2   RestingBP          918 non-null    int64  
 3   Cholesterol        918 non-null    int64  
 4   FastingBS          918 non-null    int64  
 5   MaxHR              918 non-null    int64  
 6   ExerciseAngina     918 non-null    int64  
 7   Oldpeak            918 non-null    float64
 8   HeartDisease       918 non-null    int64  
 9   TA_ChestPain       918 non-null    int32  
 10  ATA_ChestPain      918 non-null    int32  
 11  NAP_ChestPain      918 non-null    int32  
 12  ASY_ChestPain      918 non-null    int32  
 13  Normal_RestingECG  918 non-null    int32  
 14  ST_RestingECG      918 non-null    int32  
 15  LVH_RestingECG     918 non-null    int32  
 16  Up_ST_Slope        918 non

Toutes les données sont en valeur numérique. On va maintenant verifier colonne par colonne l'ecart type et la moyenne pour ressortir les valeurs abérrantes.

In [12]:
data.describe()

Age         Sex   RestingBP  Cholesterol   FastingBS  \
count  918.000000  918.000000  918.000000   918.000000  918.000000   
mean    53.510893    0.210240  132.396514   198.799564    0.233115   
std      9.432617    0.407701   18.514154   109.384145    0.423046   
min     28.000000    0.000000    0.000000     0.000000    0.000000   
25%     47.000000    0.000000  120.000000   173.250000    0.000000   
50%     54.000000    0.000000  130.000000   223.000000    0.000000   
75%     60.000000    0.000000  140.000000   267.000000    0.000000   
max     77.000000    1.000000  200.000000   603.000000    1.000000   

            MaxHR  ExerciseAngina     Oldpeak  HeartDisease  TA_ChestPain  \
count  918.000000      918.000000  918.000000    918.000000    918.000000   
mean   136.809368        0.404139    0.887364      0.553377      0.050109   
std     25.460334        0.490992    1.066570      0.497414      0.218289   
min     60.000000        0.000000   -2.600000      0.000000      0.000000   
25%    120.000000        0.000000    0.000000      0.000000      0.000000   
50%    138.000000        0.000000    0.600000      1.000000      0.000000   
75%    156.000000        1.000000    1.500000      1.000000      0.000000   
max    202.000000        1.000000    6.200000      1.000000      1.000000   

       ATA_ChestPain  NAP_ChestPain  ASY_ChestPain  Normal_RestingECG  \
count     918.000000     918.000000     918.000000         918.000000   
mean        0.188453       0.221133       0.540305           0.601307   
std         0.391287       0.415236       0.498645           0.489896   
min         0.000000       0.000000       0.000000           0.000000   
25%         0.000000       0.000000       0.000000           0.000000   
50%         0.000000       0.000000       1.000000           1.000000   
75%         0.000000       0.000000       1.000000           1.000000   
max         1.000000       1.000000       1.000000           1.000000   

       ST_RestingECG  LVH_RestingECG  Up_ST_Slope  Flat_ST_Slope  \
count     918.000000      918.000000   918.000000     918.000000   
mean        0.193900        0.204793     0.430283       0.501089   
std         0.395567        0.403770     0.495386       0.500271   
min         0.000000        0.000000     0.000000       0.000000   
25%         0.000000        0.000000     0.000000       0.000000   
50%         0.000000        0.000000     0.000000       1.000000   
75%         0.000000        0.000000     1.000000       1.000000   
max         1.000000        1.000000     1.000000       1.000000   

       Down_ST_Slope  
count     918.000000  
mean        0.068627  
std         0.252957  
min         0.000000  
25%         0.000000  
50%         0.000000  
75%         0.000000  
max         1.000000

Les colonnes a modifier sont Age, RestingBP, Cholesterol et MaxHR. 

### Nettoyage des valeures aberantes

#### RestingBP

On va maintenant supprimer les valeurs que l'on considere comme aberrantes dans la colonne de "RestingBP". Par exemple quand il est egale à zero. On considere que la personne est "morte" et donc que la ligne n'est pas pertinente. 

In [13]:
newdata = data[data['RestingBP'] == 0]
newdata.head()

Age  Sex  RestingBP  Cholesterol  FastingBS  MaxHR  ExerciseAngina  \
449   55    0          0            0          0    155               0   

     Oldpeak  HeartDisease  TA_ChestPain  ATA_ChestPain  NAP_ChestPain  \
449      1.5             1             0              0              1   

     ASY_ChestPain  Normal_RestingECG  ST_RestingECG  LVH_RestingECG  \
449              0                  1              0               0   

     Up_ST_Slope  Flat_ST_Slope  Down_ST_Slope  
449            0              1              0

Ensuite supprimer cette ligne du dataset

In [14]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 918 entries, 0 to 917
Data columns (total 19 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Age                918 non-null    int64  
 1   Sex                918 non-null    int64  
 2   RestingBP          918 non-null    int64  
 3   Cholesterol        918 non-null    int64  
 4   FastingBS          918 non-null    int64  
 5   MaxHR              918 non-null    int64  
 6   ExerciseAngina     918 non-null    int64  
 7   Oldpeak            918 non-null    float64
 8   HeartDisease       918 non-null    int64  
 9   TA_ChestPain       918 non-null    int32  
 10  ATA_ChestPain      918 non-null    int32  
 11  NAP_ChestPain      918 non-null    int32  
 12  ASY_ChestPain      918 non-null    int32  
 13  Normal_RestingECG  918 non-null    int32  
 14  ST_RestingECG      918 non-null    int32  
 15  LVH_RestingECG     918 non-null    int32  
 16  Up_ST_Slope        918 non

In [15]:
sup = 0
for index in data.index:
    i = index
    if data['RestingBP'][i] == 0:
        sup = sup + 1 
        data.drop(i, inplace=True)

print(sup,"ligne(s) a (ont) ete supprimee(s)\n")
data.info()

1 ligne(s) a (ont) ete supprimee(s)

<class 'pandas.core.frame.DataFrame'>
Index: 917 entries, 0 to 917
Data columns (total 19 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Age                917 non-null    int64  
 1   Sex                917 non-null    int64  
 2   RestingBP          917 non-null    int64  
 3   Cholesterol        917 non-null    int64  
 4   FastingBS          917 non-null    int64  
 5   MaxHR              917 non-null    int64  
 6   ExerciseAngina     917 non-null    int64  
 7   Oldpeak            917 non-null    float64
 8   HeartDisease       917 non-null    int64  
 9   TA_ChestPain       917 non-null    int32  
 10  ATA_ChestPain      917 non-null    int32  
 11  NAP_ChestPain      917 non-null    int32  
 12  ASY_ChestPain      917 non-null    int32  
 13  Normal_RestingECG  917 non-null    int32  
 14  ST_RestingECG      917 non-null    int32  
 15  LVH_RestingECG     917 non-null    int32  

#### Cholesterol

Pour le cholesterol c'est un peu plus compliqué.
On considere que le taux devient léthal à plus de 5,18 mmol/L.
Mais il existe certain cas rare ou le taux de cholesterol peut attendre des données abérrantes sans pour autant que le patient soit mort.

On parle alors hypercholestérolémie familiale. On n'est pas sur a 100% qu'un ou des patients ait cette maladie. Nous allons alors ignorer les valeures aberrantes dans la colonne "Cholesterol"

NB: on va changer les patients dont le cholestrerol est egale à zéro en la mediane de la colonne.

In [16]:
print("Nombre de valeurs égales à 0 dans Cholesterol de base :", (data['Cholesterol'] == 0).sum())

# Calculer la médiane des valeurs non nulles de cholestérol
cholesterol_median = data[data['Cholesterol'] != 0]['Cholesterol'].median()

# Remplacer les valeurs 0 par la médiane
data['Cholesterol'] = data['Cholesterol'].replace(0, cholesterol_median)

# Vérification
print("Nombre de valeurs égales à 0 dans Cholesterol :", (data['Cholesterol'] == 0).sum())
print("Médiane de Cholesterol :", data['Cholesterol'].median())

Nombre de valeurs égales à 0 dans Cholesterol de base : 171
Nombre de valeurs égales à 0 dans Cholesterol : 0
Médiane de Cholesterol : 237.0


#### Age

Pour la colonne age, on aurait pu se supprimer les valeurs aberrantes (par exemple : - 18 ans ou + 80 ans). Notre echantillon a des valeurs entre 28 ans et 77 ans. Je considere que cet echantillon est assez representatif et permettras de mieux etudies les chances de maladies cardiaques, donc je ne vais ni changer ni supprimer des valeurs.

#### MaxHR

Pour la colonne "MaxHR", qui represente la frequence cardiaque maximal c'est plus complexe également. 

On considere que le frequence cardiaque maximale doit etre comprise entre 50 et 250 bpm. Plus bas ou plus Haut, ce sont des signes de complications cardiaques certains. Heureusement dans notre dataset les données sont comprises entre 60 entre 60 et 202, donc je ne vais pas supprimer de données par rapport à cette ligne.

#### OldPeak

In [17]:
newdata = data[data['Oldpeak'] < 0]

newdata.info()

<class 'pandas.core.frame.DataFrame'>
Index: 13 entries, 321 to 512
Data columns (total 19 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Age                13 non-null     int64  
 1   Sex                13 non-null     int64  
 2   RestingBP          13 non-null     int64  
 3   Cholesterol        13 non-null     int64  
 4   FastingBS          13 non-null     int64  
 5   MaxHR              13 non-null     int64  
 6   ExerciseAngina     13 non-null     int64  
 7   Oldpeak            13 non-null     float64
 8   HeartDisease       13 non-null     int64  
 9   TA_ChestPain       13 non-null     int32  
 10  ATA_ChestPain      13 non-null     int32  
 11  NAP_ChestPain      13 non-null     int32  
 12  ASY_ChestPain      13 non-null     int32  
 13  Normal_RestingECG  13 non-null     int32  
 14  ST_RestingECG      13 non-null     int32  
 15  LVH_RestingECG     13 non-null     int32  
 16  Up_ST_Slope        13 non-null

Il y'a 13 valeures negatives de Oldpeak.

Petit contexte :

Il s'agit de la depression de ST, un valeurs numeriques calculé grace a un  électrocardiogramme (ECG). Elle correspond à une baisse du segment ST de plus de 0,5 mm en dessous de la ligne de base et peut indiquer plusieurs problèmes cardiaques.

Mais ces problèmes cardiaques peuvent être d'origine diverses. Comme Ischémie myocardique (manque d'oxygène au cœur), un effort physique intense ou encore un des Effets secondaires de médicaments. Les causes sont telement diverses que au final elles n'ont pas vraiement d'impacts sur cette études. J'ai donc décidé de conservé les 13 valeurs plus hauts.

## La colonne Heart Disease à la fin

In [18]:

data = data.assign(HeartDisease=data.pop('HeartDisease'))

## Export de la base de données nettoyés

In [19]:
data.to_csv("heart_disease_by_ceo.csv")

## Evaluation des modeles predictifs

Avec ce dataset nous allons réaliser plusieurs modelès predictifs pour determiner les chances de maladies cardiaques. Nous allons évaluer chacun de ses modèles pour determiner lequel est le plus fiable. L'évaluation se fera dans un premier temps à l'aide de la matrice de confusion et des données de précision, de rappel et de F1 puis à l'aides des critères suivants : 

- Les Performances du modèle : La précision, l'accuracy, le rappel et le F1-score. Et egalement la capacité a reconnaitre les faux positifs et les faux négatifs, pour cela nous allons utiliser les 5 mêmes patients tests pour chaque modèle.

- La robustesse du modèle : La validation croisée. Pour tout nos modèles nous utiliserons 80% des données pour l'entrainement et 20% pour le test.

- La complexité du modèle et la rapidité : Le temps d'execution.

- L'interprétabilité du modèle et son Explicabilité.